tests/slurm_plugin/test_capacity_block_manager.py

```diff
@@ -0,0 +1,745 @@
+
+    @pytest.mark.parametrize(
+        ("update_time", "expected_output"),
+        [(None, False), ("any-value", True), (datetime(2020, 1, 1, 0, 0, 0), True)],
+    )
+    def test_is_initialized(self, capacity_block_manager, update_time, expected_output):
+        capacity_block_manager._capacity_blocks_update_time = update_time
+        assert_that(capacity_block_manager._is_initialized()).is_equal_to(expected_output)
+
+    def test_ec2_client(self, capacity_block_manager, mocker):
+        ec2_mock = mocker.patch("slurm_plugin.capacity_block_manager.Ec2Client", return_value=mocker.MagicMock())
+        capacity_block_manager.ec2_client()
+        ec2_mock.assert_called_with(config="fake_boto3_config", region="eu-west-2")
+        capacity_block_manager.ec2_client()
+        ec2_mock.assert_called_once()
+
+    @pytest.mark.parametrize(
+        (
+            "is_time_to_update",
+            "previous_reserved_nodenames",
+            "capacity_blocks_from_config",
+            "capacity_blocks_info_from_ec2",
+            "nodes",
+            "slurm_reservation_creation_succeeded",
+            "expected_reserved_nodenames",
+        ),
+        [
+            (
+                # no time to update, preserve old nodenames
+                False,
+                ["node1"],
+                {},
+                [],
+                [StaticNode("queue-cb-st-compute-resource-cb-1", "ip-1", "hostname-1", "some_state", "queue-cb")],
+                [True],
+                ["node1"],
+            ),
+            (
+                # capacity block from config is empty, remove old nodenames
+                True,
+                ["node1"],
+                {},
+                [],
+                [StaticNode("queue-cb-st-compute-resource-cb-1", "ip-1", "hostname-1", "some_state", "queue-cb")],
+                [True],
+                [],  # empty because capacity block from config is empty
+            ),
+            (
+                # return first nodename because there is an internal error
+                # in the second capacity block when updating info from EC2
+                True,
+                ["node1"],
+                {"cr-123456": CapacityBlock("cr-123456", "queue-cb", "compute-resource-cb")},
+                [
+                    CapacityReservationInfo(
+                        {**FAKE_CAPACITY_BLOCK_INFO, "State": "pending", "CapacityReservationId": "cr-123456"}
+                    ),
+                    # add another id not in the capacity block to trigger an internal error
+                    # for _update_capacity_blocks_info_from_ec2, that does not stop the loop
+                    CapacityReservationInfo(
+                        {**FAKE_CAPACITY_BLOCK_INFO, "State": "active", "CapacityReservationId": "cr-234567"}
+                    ),
+                ],
+                [StaticNode("queue-cb-st-compute-resource-cb-1", "ip-1", "hostname-1", "some_state", "queue-cb")],
+                [True, True],
```

Comment:
not clear to me why if there is an internal error, the slurm reservation create should succeed

Reply:
Because it's an internal error  with the second capacity block cr-234567, that won't stop the loop. 
So the slurm reservation for the first CB will succeed.